File: src/main/java/net/yogstation/yogbot/listeners/TextCommandListener.java
package net.yogstation.yogbot.listeners;

import discord4j.core.GatewayDiscordClient;
import discord4j.core.event.domain.message.MessageCreateEvent;
import discord4j.core.object.entity.Member;
import net.yogstation.yogbot.listeners.commands.*;
import net.yogstation.yogbot.config.DiscordConfig;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.ArrayList;
import java.util.List;

@Component
public class TextCommandListener {
	private final List<TextCommand> commands;
	private final DiscordConfig config;
	
	public TextCommandListener(List<TextCommand> commands, GatewayDiscordClient client, DiscordConfig config) {
		this.commands = commands;
		this.config = config;
		
		client.on(MessageCreateEvent.class, this::handle).subscribe();
	}
	
	public List<String> getHelpMessages(Member member, boolean hidden) {
		List<String> messages = new ArrayList<>();
		for(TextCommand command : commands) {
			if(command.isHidden() != hidden) continue;
			if(command instanceof PermissionsCommand && !((PermissionsCommand) command).hasPermission(member))
				continue;
			messages.add(command.getHelpText());
		}
		return messages;
	}
	
	public Mono<?> handle(MessageCreateEvent event) {
		return Flux.fromIterable(commands)
			.filter(command -> event.getMessage().getContent().startsWith(
					config.commandPrefix +
					command.getName()
			))
			.next()
			.flatMap(command -> command.handle(event));
	}
}

File: src/main/java/net/yogstation/yogbot/listeners/TextCommandListener.kt
package net.yogstation.yogbot.listeners

import discord4j.core.GatewayDiscordClient
import discord4j.core.event.domain.message.MessageCreateEvent
import net.yogstation.yogbot.config.DiscordConfig
import net.yogstation.yogbot.listeners.commands.TextCommand
import net.yogstation.yogbot.util.DiscordUtil
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono


@Component
class TextCommandListener(
	private val commands: List<TextCommand>,
	client: GatewayDiscordClient,
	private val config: DiscordConfig
) {
	init {
		client.on(MessageCreateEvent::class.java) { event: MessageCreateEvent -> handle(event) }.subscribe()
	}

	fun handle(event: MessageCreateEvent): Mono<*> {
		if(!event.message.content.startsWith(config.commandPrefix)) return Mono.empty<Any>()
		val command: TextCommand = commands.firstOrNull { command: TextCommand ->
			event.message.content.startsWith(
				config.commandPrefix + command.name
			)
		} ?: return DiscordUtil.reply(event, "Command ${event.message.content.split(" ", limit = 2)[0]} not found")
		return command.handle(event)
	}
}
